singer_sdk/typing.py

```diff
@@ -159,25 +171,87 @@ def set_defaults(
     )
 
 
-class JSONTypeHelper:
+class DefaultInstanceProperty:
+    """Property of default instance.
+
+    Descriptor similar to ``property`` that decorates an instance method to retrieve
+    a property from the instance initialized with default parameters, if the called on
+    the class.
+    """
+
+    def __init__(self, fget: Callable) -> None:
+        """Initialize the decorator.
+
+        Args:
+            fget: The function to decorate.
+        """
+        self.fget = fget
+
+    def __get__(self, instance: P, owner: type[P]) -> Any:  # noqa: ANN401
+        """Get the property value.
+
+        Args:
+            instance: The instance to get the property value from.
+            owner: The class to get the property value from.
+
+        Returns:
+            The property value.
+        """
+        if instance is None:
+            instance = owner()
```

Comment:
@aaronsteers I'd like to get your thoughts on this approach. It replaces the `classproperty` descriptor with one that gets the attribute from

1) the instance of the class, e.g. `StringType().type_dict` or
2) an instance initialized with no arguments, e.g. `StringType.type_dict` (under the hood ``StringType().type_dict` is called).

also cc @WillDaSilva if you think this is not a good pattern

Reply:
@edgarrmondragon Let me know if I understand the context correctly.

We currently use `classproperty` with the expectation that the class object will be used directly instead of instances of the class.

We want to allow customization by providing `allowed_values` and `examples` to the class, but we don't want to introduce a breaking change by requiring the class be initialized in all cases.

To accommodate this new feature while continuing to support the old way to using the classes directly (instead of via an instance) this change adds a new property type that uses the instance if it exists, and creates an instance (with the default keyword arguments) otherwise.

Seems reasonable. One hazard to watch out for is that none of the classes can ever have a parameter without a default. We _could_ add a metaclass to `JSONTypeHelper` that enforces that restriction so that we never accidentally forget and introduce a bug, but that may be overkill since our tests should always catch bugs like that.

Reply:
That's a very accurate description of the problem @WillDaSilva. I agree that we risk declaring a class with positional arguments but we don't often add new type helper classes and you're right that tests should catch them.

I've added more doctest for other type helper classes.